File: Minjae/Minjae/정렬/국영수.swift
//
//  국영수.swift
//  Minjae
//
//  Created by 김민재 on 2023/02/05.
//

import Foundation

// N이 최대 100,000까지 주어지기 때문에
// O(NlogN)의 알고리즘으로 설계하면 풀릴 것
// Swift에서의 sort은 O(n log n)을 보장해준다.
let N = Int(readLine()!)! // N명
var arr: [(String, Int, Int, Int)] = []
for _ in 0..<N {
    let input = readLine()!.split(separator: " ")
    let name = String(input[0])
    let korean = Int(input[1])!
    let english = Int(input[2])!
    let math = Int(input[3])!
    
    arr.append((name, korean, english, math)) // 튜플 형식으로 저장
}

arr.sort { a, b in // 조건에 맞게 정렬하기
    if a.1 == b.1 {
        if a.2 == b.2 {
            if a.3 == b.3 {
                return a.0 < b.0
            }
            return a.3 > b.3
        }
        return a.2 < b.2
    }
    return a.1 > b.1
}

for a in arr { // 이름만 출력하기
    print(a.0)
}

File: jejune/src/정렬/국영수.java
package jejune.src.정렬;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

public class 국영수 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        Student[] students = new Student[n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            students[i] = new Student(
                    st.nextToken(),
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken()),
                    Integer.parseInt(st.nextToken())
            );
        }


        Arrays.sort(students, new Comparator<Student>() {
            @Override
            public int compare(Student o1, Student o2) {
                if (o1.korean < o2.korean) {
                    return 1;
                } else if (o1.korean == o2.korean) {
                    if (o1.english > o2.english) {
                        return 1;
                    } else if (o1.english == o2.english) {
                        if (o1.math < o2.math) {
                            return 1;
                        } else if (o1.math == o2.math) {
                            return o1.name.compareTo(o2.name);
                        }
                    }
                }
                return -1;
            }
        });

        for (Student student : students) {
            System.out.println(student.name);
        }
    }
}

class Student {
    String name;
    int korean;
    int english;
    int math;

    Student(String name, int korean, int english, int math) {
        this.name = name;
        this.korean = korean;
        this.english = english;
        this.math = math;
    }
}
